File: backend/src/main/java/com/matheus/book/application/customer/service/CustomerService.java
package com.matheus.book.application.customer.service;

import com.matheus.book.application.book.dto.response.BookSimpleResponse;
import com.matheus.book.application.customer.dto.request.CreateCustomerRequest;
import com.matheus.book.application.customer.dto.request.EditCustomerRequest;
import com.matheus.book.application.customer.dto.response.CustomerResponse;
import com.matheus.book.application.customer.enums.CustomerFilter;
import com.matheus.book.domain.customer.entity.Customer;
import com.matheus.book.domain.customer.repository.CustomerRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Service
@RequiredArgsConstructor
public class CustomerService {

    private final CustomerRepository customerRepository;

    public CustomerResponse create(CreateCustomerRequest request){
        Customer customer = new Customer(request.name());

        customerRepository.save(customer);

        return toResponse(customer);
    }

    public Page<CustomerResponse> findAllByName(String name, CustomerFilter filter,
                                                int page, int size){
        Page<Customer> customers;

        Pageable pageable = PageRequest.of(
                page,
                size,
                Sort.by(
                        Sort.Order.asc("name").ignoreCase()
                )
        );

        switch (filter){
            case ALL -> customers = customerRepository.findAllByName(name, pageable);
            case WITH_BORROWED_BOOKS ->
                    customers = customerRepository.findWithBorrowedBooksByName(name, pageable);

            case WITHOUT_BORROWED_BOOKS ->
                    customers = customerRepository.findWithoutBorrowedBooksByName(name, pageable);

            default -> customers = customerRepository.findAll(pageable);
        }

        return customers.map(this::toResponse);
    }

    public CustomerResponse findById(Long id){
        Customer customer = findCustomerById(id);

        return toResponse(customer);
    }

    public int info(){
        return customerRepository.findAll().size();
    }

    @Transactional
    public void editInfo(Long id, EditCustomerRequest request){
        Customer customer = findCustomerById(id);

        customer.edit(request.name());
    }

    @Transactional
    public void delete(Long id){
        Customer customer = findCustomerById(id);

        if(!customer.getBooks().isEmpty()){
            throw new ResponseStatusException(HttpStatus.CONFLICT,
                    "This customer cannot be deleted because they have books");
        }

        customerRepository.delete(customer);
    }

    private Customer findCustomerById(Long id){
        return customerRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.NOT_FOUND, "Customer not found"));
    }

    private CustomerResponse toResponse(Customer customer){
        return new CustomerResponse(
                customer.getId(),
                customer.getName(),
                customer.getBooks()
                        .stream()
                        .map((book) -> new BookSimpleResponse(
                                book.getId(),
                                book.getName(),
                                book.getPublisher().getName(),
                                book.isAvailable()
                        )).toList()
        );
    }

}

File: frontend/src/app/home/api/customer/service/customer.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { CustomerRequest } from '../dto/request/customer-request';
import { CustomerResponse } from '../dto/response/customer-response';
import { Observable } from 'rxjs';
import { PageResponse } from '../../page/page-response';

@Injectable({
  providedIn: 'root',
})
export class CustomerService {
  private readonly API_URL = '/api/customers'

  constructor(private http: HttpClient){}

  create(request: CustomerRequest): Observable<CustomerResponse>{
    return this.http.post<CustomerResponse>(`${this.API_URL}`, {...request});
  }

  findAllByName(name: string, filter: string, page: number): Observable<PageResponse<CustomerResponse>>{
    return this.http.get<PageResponse<CustomerResponse>>(`${this.API_URL}`, { 
      params: { name, filter, page}
    });
  }

  findById(id: number): Observable<CustomerResponse>{
    return this.http.get<CustomerResponse>(`${this.API_URL}/${id}`)
  }

  info(): Observable<number>{
    return this.http.get<number>(`${this.API_URL}/info`);
  }

  editInfo(id: number, name: string): Observable<void>{
    return this.http.patch<void>(`${this.API_URL}/${id}`, { name })
  }

  delete(id: number): Observable<void>{
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }
}
